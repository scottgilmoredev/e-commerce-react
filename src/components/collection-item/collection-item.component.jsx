import React from 'react';
import { connect } from 'react-redux';

// Actions
import { addItem } from '../../redux/cart/cart.actions';

// Styles
import {
    AddButton,
    BackgroundImage,
    CollectionFooterContainer,
    CollectionItemContainer,
    NameContainer,
    PriceContainer,
  } from './collection-item.styles';

/**
 * Display a collection item.
 * @param {Function} addItem - add item action.
 * @param {Object} item - id, imageUrl, name, price.
 */
export const CollectionItem = ({ addItem, item }) => {
    const { imageUrl, name, price } = item;

    return (
        <CollectionItemContainer>
            { /* Item image */ }
            <BackgroundImage
                className='image'
                imageUrl={ imageUrl }   
            />

            { /* Item name / price */ }
            <CollectionFooterContainer>
                <NameContainer>{ name }</NameContainer>
                <PriceContainer>{ price }</PriceContainer>
            </CollectionFooterContainer>

            { /* Add to cart */ }
            <AddButton
                inverted
                onClick={ () => addItem(item) }
            >
                Add to cart
            </AddButton>
        </CollectionItemContainer>
    )
};

const mapDispatchToProps = dispatch => ({
    addItem: item => dispatch(addItem(item)),
});

export default connect(null, mapDispatchToProps)(CollectionItem);
