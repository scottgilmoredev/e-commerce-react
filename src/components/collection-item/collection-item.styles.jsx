import styled from 'styled-components';
import CustomButton from '../custom-button/custom-button.component';

export const CollectionItemContainer = styled.div`
  align-items: center;
  display: flex;
  flex-direction: column;
  height: 350px;
  position: relative;
  width: 22vw;

  &:hover {
    .image {
      opacity: 0.8;
    }

    button {
      display: flex;
      opacity: 0.85;
    }
  }

  @media screen and (max-width: 800px) {
    width: 40vw;

    &:hover {
      .image {
        opacity: unset;
      }

      button {
        opacity: unset;
      }
    }
  }
`;

export const AddButton = styled(CustomButton)`
  display: none;
  opacity: 0.7;
  position: absolute;
  top: 255px;
  width: 80%;
  height: 40px;

  @media screen and (max-width: 800px) {
    display: flex;
    opacity: 0.9;
    min-width: unset;
    padding: 0 10px;
  }
`;

export const BackgroundImage = styled.div`
  background-image: ${({ imageUrl }) => `url(${imageUrl})`};
  background-position: center;
  background-size: cover;
  height: 95%;
  margin-bottom: 5px;
  width: 100%;
`;

export const CollectionFooterContainer = styled.div`
  display: flex;
  font-size: 18px;
  height: 5%;
  justify-content: space-between;
  width: 100%;
`;

export const NameContainer = styled.span`
  margin-bottom: 15px;
  width: 90%;
`;

export const PriceContainer = styled.span`
  text-align: right;
  width: 10%;
`;

// Display names.
CollectionItemContainer.displayName = 'CollectionItemContainer';
AddButton.displayName = 'AddButton';
BackgroundImage.displayName = 'BackgroundImage';
CollectionFooterContainer.displayName = 'CollectionFooterContainer';
NameContainer.displayName = 'NameContainer';
PriceContainer.displayName = 'PriceContainer';
