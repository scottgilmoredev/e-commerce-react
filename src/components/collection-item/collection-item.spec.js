import React from 'react';
import { shallow } from 'enzyme';

// Component
import { CollectionItem } from './collection-item.component';

describe('CollectionItem component', () => {
    let wrapper;
    let mockAddItem;
    const mockImageUrl = 'www.testImage.com';
    const mockName = 'black hat';
    const mockPrice = 10;

    beforeEach(() => {
        mockAddItem = jest.fn();

        const mockProps = {
            item: {
                imageUrl: mockImageUrl,
                name: mockName,
                price: mockPrice,
            },
            addItem: mockAddItem,
        };

        wrapper = shallow(<CollectionItem { ...mockProps } />);
    });

    // Render the component.
    it('should render the CollectionItem component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    // Add item.
    it('should call addItem when AddButton is clicked', () => {
        wrapper.find('AddButton').simulate('click');
        expect(mockAddItem).toHaveBeenCalled();
    });

    // Render image url.
    it('should render the imageUrl as a prop on BackgroundImage', () => {
        expect(wrapper.find('BackgroundImage').prop('imageUrl')).toBe(mockImageUrl);
    });

    // Render name.
    it('should render the name prop in NameContainer', () => {
        expect(wrapper.find('NameContainer').text()).toBe(mockName);
    });

    // Render price.
    it('should render the price prop in PriceContainer', () => {
        const price = parseInt(wrapper.find('PriceContainer').text());
        expect(price).toBe(mockPrice);
    });
});
