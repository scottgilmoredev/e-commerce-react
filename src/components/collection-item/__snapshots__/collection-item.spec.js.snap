// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionItem component should render the CollectionItem component 1`] = `ShallowWrapper {}`;
